[AAP-46090] Establish contribution guidelines (#167)
ansible/terraform-provider-aap: CONTRIBUTING.md

# Contributing to terraform-provider-aap

> 

Reviewer: I am missing some information on "what to work on" and the expectations on which contributions will be accepted.

We don't want people to invest time in stuff we are not interested in merging. I suggest highligting a remark about first opening an Issue, receive feedback, and only when something is agreed on, open the PR.

And clearly state that PRs that have not gone through this process might not be merged ever.

# Contributing to terraform-provider-aap

Thank you for contributing to the Ansible Automation Platform Terraform Provider!

> These guidelines document the standards the Inframesh team follows and holds each other accountable to. Whether you're an internal team member or external contributor, following these guidelines ensures consistent, high-quality contributions.

Reviewer 2: ```suggestion
These guidelines document the standards for contribution. Whether you're an internal team member or external contributor, following these guidelines ensures consistent, high-quality contributions.
```

Please remove specific references to our specific internal structures/teams/resources in public repos. External contributors don't need this information and internal contributors learn it through internal channels.

## Quick Start

| Step | Command | Purpose |
|------|---------|---------|
| 1. Fork & clone | `git clone https://github.com/<your-username>/terraform-provider-aap` | Get the code |

> | 2. Choose base branch | `git checkout main` or `git checkout dev-preview` | See [Branch Strategy](#branch-strategy) |

Reviewer 2: Branches should be created from `main`, see https://github.com/ansible/terraform-provider-aap/blob/main/README.md#branching-model.

| CI must pass | Automatic | ✅ Yes |

**Requirements**: 
- All PRs must include unit tests for new code
- New code must have minimum 80% test coverage ([SonarCloud](https://sonarcloud.io/project/overview?id=ansible_terraform-provider-aap))

Run `make testcov` to generate coverage report locally at `./unit-testing.cov`.

### Acceptance Tests

| Action | Command | Required? |
|--------|---------|-----------|
| Setup test resources | `ansible-playbook testing/playbook.yml` | First time only |
| Source test vars | `source testing/acceptance_test_vars.env` | Before each run |
| Run acceptance tests | `make testacc` | ✅ Yes |

**Requirement**: PRs must include acceptance test output in PR description.

#### Environment Setup

```bash
# Set AAP credentials
export AAP_USERNAME=admin
export AAP_PASSWORD=your_password
export AAP_HOSTNAME=http://localhost:9080
export AAP_INSECURE_SKIP_VERIFY=true
export AAP_VALIDATE_CERTS=false

# For AAP 2.4
export CONTROLLER_OPTIONAL_API_URLPATTERN_PREFIX="/api/"

# Install ansible collection
ansible-galaxy collection install -r testing/requirements.yml

# Create test resources
ansible-playbook testing/playbook.yml
```

> ⚠️ **WARNING**: Acceptance tests modify real AAP resources. Use a dedicated test instance.

Reviewer 2: This content also exists in the README. Instead of duplicating it (and having a second place to maintain it), can we reference the existing steps?

## Code Standards

### Pull Request Checklist

- [ ] Unit tests added/updated
- [ ] Code coverage ≥80% for new code (`make testcov`)
- [ ] Acceptance tests pass (include output in PR)
- [ ] Linting passes (`make lint`)
- [ ] Documentation updated (if applicable)
- [ ] Changelog entry added (if applicable)
- [ ] Examples updated (if applicable)
- [ ] AI attribution noted (if substantial AI assistance used)
- [ ] PR is focused on single feature/fix

### Focused PRs

**Keep PRs focused on one feature or fix.** If you find yourself making multiple unrelated changes, split them into separate PRs.

**Good examples:**
- ✅ New resource + related changes to same resource
- ✅ Bug fix across multiple related resources
- ✅ Documentation updates for one feature

**Should be split:**
- ❌ New feature + unrelated bug fixes
- ❌ Changes to multiple unrelated resources
- ❌ Refactoring + new features

**Why?** Focused PRs are easier to review, test, and revert if needed.

### New Resources/Data Sources

| Requirement | Location | Notes |
|-------------|----------|-------|
| Implementation | `internal/provider/` | Use Plugin Framework |
| Unit tests | `*_test.go` | Test all CRUD operations |

> | Acceptance test | `*_test.go` | Use `TF_ACC=1` |

Reviewer 2: I think we could use more detail in the descriptions here. Unit tests will need to do more than CRUD to get to 80% and there's more to acceptance tests than `TF_ACC=1`

## Review Process

| Stage | Timeline | Action |
|-------|----------|--------|
| PR opened | Within 2 business days | Initial review |
| Changes requested | N/A | Address feedback |
| Approved | Within 1 business day | Merge to main |

### What Reviewers Check

- [ ] Tests pass and cover new code
- [ ] Code coverage ≥80% for new code
- [ ] Linting passes
- [ ] Documentation is clear and complete
- [ ] Code follows existing patterns
- [ ] PR scope is focused
- [ ] Changelog entry (if needed)

## AI Code Assistant Usage

If you use AI code assistants, please follow these guidelines:

| Guideline | Details |
|-----------|---------|
| **Human verification** | Always review, test, and understand AI-generated code before submitting |
| **When to attribute** | Note substantial AI assistance in PR description using `Assisted-by:` |
| **Security** | Never input sensitive data (API keys, credentials, customer data) into AI tools |
| **Code quality** | AI-generated code must meet the same standards (tests, coverage, linting) |
| **License compliance** | Ensure AI suggestions don't introduce incompatible licenses |

**Example PR description:**
```
Implemented retry logic for API calls

> Assisted-by: GitHub Copilot

Reviewer 2: Good notes. We should also include any assitance attribution in the commit messages.

**Important:** Pull requests submitted without prior discussion may not be merged, even if the code is high quality. This helps ensure your time is well spent on contributions we can accept.

## Quick Start

| Step | Command | Purpose |
|------|---------|---------|
| 1. Fork & clone | `git clone https://github.com/<your-username>/terraform-provider-aap` | Get the code |
| 2. Create branch from main | `git checkout main && git checkout -b feature-name` | Isolate your work |
| 3. Make changes | Edit code, add tests | Implement feature |
| 4. Run checks | `make lint && make test && make testacc` | Verify quality |
| 5. Commit | `git commit -m "description"` | Save changes |
| 6. Push | `git push origin feature-name` | Share work |
| 7. Open PR | Create pull request on GitHub | Request review |

For branching strategy details, see the [README](README.md#branching-model).

## Requirements

| Tool | Version | Installation Guide |
|------|---------|-------------------|
| Go | 1.24 | <https://go.dev/doc/install> |
| Terraform | 1.0+ | <https://developer.hashicorp.com/terraform/install> |
| golangci-lint | 2.5.0 | <https://golangci-lint.run/usage/install/> |

> | AAP/AWX instance | 2.4+ | <https://github.com/ansible/awx/blob/devel/INSTALL.md> |

Reviewer 2: I saw that @tyraziel requested specific versions here but I disagree and have different feedback.

This information is already captured in the README: https://github.com/ansible/terraform-provider-aap?tab=readme-ov-file#requirements. Instead of duplicating that information and recording specific versions numbers (that will be out of date unless we commit to regularly updating these docs), let's just tell people where to find the information about what tools and versions are needed.

Author: Done.
